### PR TITLE
Remove armeria-brave module

### PR DESCRIPTION
**Pull request by codefromthecrypt**
Motivation:

armeria-brave was changed in a unique way in 1.27.1 to pin an old version of the library, even though there are no code compatibility issues.

What happened was some folks at LINE have a metrics library, and instead of depending on zipkin-reporter, they relied on brave's indirect dependency version. In a rush to fix this quick, brave-armeria was reverted to 5.x and a new module was made for brave6. Another was also made for brave5 despite almost certainly no more released on that line of code.

The problem is that a normal person would expect "armeria-brave" to have the more recent version of brave, and instead they will be pinned and not receive code updates. They will have no easy way to find this out because there is no code in "armeria-brave" to mark deprecated. This basically means people will unknowingly always stay at an old version.

Another choice would have been to still make the separate modules, but make the main "armeria-brave" tactically set the old dependency it had before. This is better because it allows a migration without serious impact. This is what I believe would have been the choice if it wasn't rushed to stop a fire at LINE.

I hope people can consider this and merge it with any modifications that keep LINE not having to manage versions, yet not pin the rest of the world to an old library.

If not, I hope people can be transparent with who is making the decision and literally what the problem is, as this doesn't seem difficult to fix from how it was described in #5438. Maybe others who know the library like @kojilin can explain if somehow managing a dependency like this uniquely cannot work at LINE. I often promote armeria as a project for everyone, not just LINE, and hope we can turn this around.

Modifications:

- Remove `armeria-brave` module

Result:

- (Breaking) The `armeria-brave` module has been fully removed. Please switch to using `armeria-brave6` along with additional dependencies:
  - `zipkin-reporter` if you are utilizing `AsyncZipkinSpanHandler`.
  - `zipkin` if you are utilizing `ZipkinSpanHandler`.
  - Alternatively, use `armeria-brave5` if you wish to retain the previous module functionality.

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
fyi I published this local and here are the differences, assuming you don't use the armeria-bom and just look at indirect deps from armeria-brave. I tried to match the pre-reporter 3.0 thing, even though brave5 uses reporter 3.

com.linecorp.armeria:armeria-brave:1.26.4
```
[INFO]    +- io.zipkin.brave:brave:jar:5.16.0:compile
[INFO]    |  \- io.zipkin.reporter2:zipkin-reporter-brave:jar:2.16.3:compile
[INFO]    |     \- io.zipkin.reporter2:zipkin-reporter:jar:2.16.3:compile
[INFO]    |        \- io.zipkin.zipkin2:zipkin:jar:2.23.2:compile
[INFO]    +- io.zipkin.brave:brave-instrumentation-http:jar:5.16.0:compile
```

com.linecorp.armeria:armeria-brave:<this PR>
```
[INFO]    |  +- io.zipkin.brave:brave:jar:6.0.0:compile
[INFO]    |  \- io.zipkin.brave:brave-instrumentation-http:jar:6.0.0:compile
[INFO]    +- io.zipkin.reporter2:zipkin-reporter-brave:jar:2.17.2:compile
[INFO]    |  \- io.zipkin.reporter2:zipkin-reporter:jar:2.17.2:compile
[INFO]    |     \- io.zipkin.zipkin2:zipkin:jar:2.27.0:compile
```

**Review by jrhee17 — APPROVED** [review]
Changes make sense to me 👍

**Comment by codefromthecrypt** [comment]
Note: if 1.27.1 works at line than we actually don't need to pin to reporter 2.x. Basically if 1.27.1 works, it could be as simple as adding a zipkin core jar dep.

As you can see 1.27.1 armeria-brave uses a recent zipkin-reporter version:

```
[INFO]    +- com.linecorp.armeria:armeria-brave5:jar:1.27.1:compile
[INFO]    |  +- io.zipkin.brave:brave:jar:5.18.0:compile
[INFO]    |  |  +- io.zipkin.reporter2:zipkin-reporter-brave:jar:3.0.0:compile
[INFO]    |  |  |  \- io.zipkin.reporter2:zipkin-reporter:jar:3.0.0:compile
[INFO]    |  |  \- io.zipkin.zipkin2:zipkin:jar:2.27.0:compile
[INFO]    |  \- io.zipkin.brave:brave-instrumentation-http:jar:5.18.0:compile
```

TL;DR; is if in fact 1.27.1 fixed it, maybe we can get by with most recent zipkin-reporter 3.x, but an extra dep on `io.zipkin.zipkin2:zipkin` and a note about why.

**Comment by codefromthecrypt** [comment]
SUMMARY:

My suspicion is that this is more about a "convenience dep" on `io.zipkin.zipkin2:zipkin` rather than specifically about reporter 2.x vs 3.x. If it was about zipkin-reporter 2.x vs 3.x, the armeria 1.27.1 would have failed to work.

So, I think we should update this to the latest reporter and also add the "convenience dep" this makes version maintenance a lot easier as you can blindly update these versions and not have to worry about 2.x or 3.x

**Comment by codefromthecrypt** [comment]
@jrhee17 is there any way to verify this at LINE with one of the apps that struggled or a similar reproducer? I want to make sure we get this right..

**Comment by jrhee17** [comment]
> So, I think we should update this to the latest reporter and also add the "convenience dep" this makes version maintenance a lot easier as you can blindly update these versions and not have to worry about 2.x or 3.x

I didn't have time to actually analyze the dependency relationship between brave, zipkin, armeria so let me know if I'm misunderstanding anything 😄 

The report we received was literally `NoClassDefFoundError: zipkin2/reporter/brave/ZipkinSpanHandler`.
Rather than pinning to a major version (2.x or 3.x), I think it makes sense to follow compatibility between zipkin and brave.

So if brave6 compatibility with zipkin3 is actively tested, I think it probably makes sense to update to the latest version

>  but an extra dep on io.zipkin.zipkin2:zipkin and a note about why.

If we directly use zipkin 3, then couldn't we drop zipkin 2?

> is there any way to verify this at LINE with one of the apps that struggled or a similar reproducer

Once we merge this to main, I can request one of our users to test with a snapshot version

**Comment by codefromthecrypt** [comment]
@jrhee17 so the thing is that it is possible to use zipkin2/reporter/brave/ZipkinSpanHandler without even the zipkin core library present. Because the input type is brave's MutableSpan.

So, in this case I think if you can test after merge to master, let's not include the zipkin core library at all, just add the dep on zipkin-reporter-brave. If things work, that's better as less to manage and less dependencies. For example, zipkin's core jar is a few hundred KB.

**Comment by codefromthecrypt** [comment]
actually zipkin2/reporter/brave/ZipkinSpanHandler needs zipkin core jar, I can see it does. The async one does not. I'll keep the zipkin core dep.

**Comment by codefromthecrypt** [comment]
ok I set current versions and also added a missing test about what this is all about, basically that there are no class problems in indirect deps

File: brave/brave/src/test/java/com/linecorp/armeria/brave/CompilationTest.java
Comment (codefromthecrypt) [inline brave/brave/src/test/java/com/linecorp/armeria/brave/CompilationTest.java:50]: so this test should give 100pct confidence that a dep change won't break without knowing

**Comment by codefromthecrypt** [comment]
squashed

**Comment by ikhoon** [comment]
The changes look acceptable since there are no code breaking changes in Brave 6.
@mauhiz Any thoughts on this?

**Comment by mauhiz** [comment]
I'm good with this. The PR submitter took extra care to maintain
compatibility, and now that he is a code owner for the module he will be
able to ensure that users won't be left wishing they didn't upgrade :)

File: brave/brave/src/test/java/com/linecorp/armeria/brave/CompilationTest.java
Comment (minwoox) [inline brave/brave/src/test/java/com/linecorp/armeria/brave/CompilationTest.java:41]: nit:
```suggestion
class BraveCompilationTest {
```

File: brave/brave/build.gradle
Comment (minwoox) [inline brave/brave/build.gradle:9]: Shouldn't we add this to brave6 module instead?

**Review by minwoox — COMMENTED** [review]
What are your thoughts on completely removing the armeria-brave module and instead providing armeria-brave5 and armeria-brave6, similar to how we handle the Jetty module? I believe this would eliminate any ambiguity.

File: brave/brave/build.gradle
Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: `AsyncZipkinSpanHandler` is the only good way to report spans, `ZipkinSpanHandler` was a legacy bridge, and pins zipkin classes (rather than brave's native MutablSpan which has its own codecs)

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: starting in brave 5, it didn't need a dependency on the library zipkin server needed. brave 6 removed the deprecated functions that still accessed those types. brave 6 can work with zipkin-reporter 2 (which uses zipkin core types) or zipkin-reporter 3 (which doesn't). If we pin brave 6 here to zipkin core types it adds a dependency that it several hundred K and won't be used. I highly suggest the LINE internal code stop using zipkin core types and use MutableSpan instead, or declare their own dependency on core types.

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: I was told this was about metrics, so it is far more efficient for the internal code to skip entirely switching to zipkin core types. here's an example of how to do something like thathttps://github.com/openzipkin/brave/blob/master/brave/src/test/java/brave/features/handler/SpanMetricsCustomizer.java

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: also, I will help port it if someone puts it public, really we don't want to pin deps for an edge case, especially as it took a lot of effort to remove this dep ;)

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: In general I'm very curious about what armeria's expectations are about indirect dependencies.. that a project should forever declare dependencies on anything they depended on at some point, even after switching major version numbers?

I think at some point people should start to declare dependencies on the types they rely on in main code, that are indirectly supplied by another artifact. Is this strange? How does netty handle it? Has moving from 3 to 4 included every indirect dependencies 3 ever used?

Comment (minwoox) [reply brave/brave/build.gradle:9]: `armeria-brave` was was simply retained as an alias of `armeria-brave5` so I thought adding extra dependency declaration to the `build.gradle` doesn't make sense.
Have you thought about this comment?
https://github.com/line/armeria/pull/5453#pullrequestreview-1879249321

Comment (minwoox) [reply brave/brave/build.gradle:9]: > In general I'm very curious about what armeria's expectations are about indirect dependencies.. that a project should forever declare dependencies on anything they depended on at some point, even after switching major version numbers?

We don't usually declare dependencies unless there's a version conflict issue. ;) Again, I believe we just wanted to keep the `armeria-brave` module as an alias of `armeria-brave5` (now it's for `armeria-brave6`), so I thought it didn't make sense to add them. 😆

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: I think armeria-brave5 is dead, as there's a near zero chance of any release on that version. Because the problem is about an indirect dep and LINE is already on armeria-brave (which is based on 6), I think removing both and keeping only brave6 is the best course, although it is a shame it can't just be armeria-brave

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: put another way, while I added tests to ensure armeria-brave can compile random types also in zipkin, I feel it is wrong. I'm only doing this in deference to the major sponsor of this project. If keeping armeria-brave with the extra deps helps fund armeria. practical enough. OTOH it isn't needed.

What is important is to not mess up what's currently called brave6 by adding random deps to it which can re-introduce conflicts. If we do anyway, for example to get just one armeria-brave, it is fine, just let's put a comment loudly on when this will be removed and that its purpose is to help internal code move over to adding an explicit dep (even if via zipkin-reporter-bom) on types they use.

Comment (minwoox) [reply brave/brave/build.gradle:9]: > I think removing both and keeping only brave6 is the best course, although it is a shame it can't just be armeria-brave

Okay, let me talk to the team tomorrow once more if we can completely remove `armeria-brave5` or not. If it's not possible, you still want to retain `armeria-brave` module (with brave6). Is that correct?

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: so after shower, here's an idea.

we *add an armeria sender* somewhere, and in that case someone can depend on that and get the correct recent reporter deps. If we add it to armeria-brave, basically that's fine and at least isn't random to add reporter (v3) anymore (because [sender](https://github.com/openzipkin/brave-example/blob/master/armeria/src/main/java/brave/example/WebClientSender.java) and [its tests](https://github.com/openzipkin/zipkin-reporter-java/tree/master/urlconnection/src/test/java/zipkin2/reporter/urlconnection) will need zipkin-reporter 3 types).

What we are left with is that large zipkin dep `io.zipkin.zipkin2:zipkin` which this was all about. zipkin-reporter-brave and all the senders exclude it because it is  huge and unused even when reporting to zipkin, if you use the most current way `AsyncZipkinSpanHandler` as our [armeria-example](https://github.com/openzipkin/brave-example/blob/master/armeria/src/main/java/brave/example/HttpTracingFactory.java) does.
```xml
    <dependency>
      <groupId>${project.groupId}</groupId>
      <artifactId>zipkin-reporter</artifactId>
      <version>${project.version}</version>
      <!-- Senders don't use zipkin types. Excluding allows brave users to
           avoid them by default. -->
      <exclusions>
        <exclusion>
          <groupId>io.zipkin.zipkin2</groupId>
          <artifactId>zipkin</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

So, basically it boils things down to one thing.. do we need to keep adding a dep on this? If so, we need to manage zipkin's version, and keep bumping like dependabot even though it isn't used, and also any app that uses brave will be a few hundred K larger than it needs to be.

That's all the info and pros and cons, so I'll leave you to decison making. Thanks for listening @minwoox!

Comment (minwoox) [reply brave/brave/build.gradle:9]: Let me answer your question.

> In general I'm very curious about what armeria's expectations are about indirect dependencies.. that a project should forever declare dependencies on anything they depended on at some point, even after switching major version numbers?

It depends. 😉 If the indirect dependencies are tightly coupled in terms of using our API, we want to retain them even after a major version update so that: 1) users can seamlessly upgrade the Armeria version without modifying their dependency configuration, and 2) we can manage the versions of indirect dependencies. If they're not coupled, we simply let users choose whatever they prefer.

> do we need to keep adding a dep on this? If so, we need to manage zipkin's version, and keep bumping like dependabot even though it isn't used,

So it's the same context here. If most of `armeria-brave` users use `AsyncZipkinSpanHandler`, we definitely need to add `zipkin-reporter`. Let me investigate how armeria-brave users use the API so that we can make a decision. 😉

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: What's probably confusing is that in Brave 4 these were coupled, Brave 5 decoupled them but kept back compat, and Brave 6 removed coupling.

Adding zipkin-reporter-brave is fine, as it is small. The side effect of adding this is that we have to pin a specific version of zipkin-reporter (3.x now), and will have a problem again on 4.x. zipkin-reporter is a different library than brave, even if it has a bridge. For example, it is also used in otel.

regardless,  it won't bring in the polarizing and unneeded dep io.zipkin.zipkin2 dep.

So there are two decisions
 1. for convenience, pin reporter (as done now in armeria-brave) and know this needs to be revisited in the event of a reporter 4?
 2. if you do that, also add the pretty large io.zipkin2:zipkin dep

over to you!

Comment (minwoox) [reply brave/brave/build.gradle:9]: After discussing with the team, we've decided that removing the armeria-brave module is the best approach to address the indirect dependency problem and eliminate any ambiguity by maintaining three different modules.

Previously, we retained the armeria-brave module to avoid causing breaking changes for users. However, we've now agreed that removing the armeria-brave module and providing armeria-brave5 is acceptable, as users will not need to modify their code. They only need to change the module name in the build file.

By removing armeria-brave, which has dependencies on zipkin, we can resolve the indirect dependency issue with armeria-brave6, which only includes the brave dependency.

Furthermore, we plan to deprecate the `armeria-brave5` module by adding `@Deprecated` to the `package-info.java` file within the module. This will facilitate users in migrating to `armeria-brave6`.

Please, let me know if we have any missing points and what you think!
Thanks, Adrian!

Comment (codefromthecrypt) [reply brave/brave/build.gradle:9]: sounds like a good plan, I didn't think about package-info.java for deprecation.. that's a neat trick I'll add to my bag 👍

**Comment by codefromthecrypt** [comment]
fyi new toys for those using the deps in question https://github.com/openzipkin/zipkin-reporter-java/releases/tag/3.3.0

**Comment by codefromthecrypt** [comment]
so I assume we don't need this PR anymore as basically the change is removing armeria-brave and adding the package-info to deprecate brave5?

**Comment by minwoox** [comment]
> so I assume we don't need this PR anymore as basically the change is removing armeria-brave and adding the package-info to deprecate brave5?

Yep. Let me push the patch to this PR. 😉

File: dependencies.toml
Comment (codefromthecrypt) [inline dependencies.toml:163]: since bumping, might as well!

```suggestion
zipkin-reporter = "3.3.0"
```

Comment (minwoox) [reply dependencies.toml:163]: At let me remove these dependencies since they are not used. 😉

**Review by minwoox — APPROVED** [review]
Thanks, Adrian!

File: brave/brave5/src/main/java/com/linecorp/armeria/client/brave/package-info.java
Comment (kojilin) [inline brave/brave5/src/main/java/com/linecorp/armeria/client/brave/package-info.java:2]: using 2024 for new file?

File: brave/brave5/build.gradle
Comment (kojilin) [inline brave/brave5/build.gradle:17]: curious do we prefer to use the convention like
```
def generateSources =  tasks.register('generateSources', Copy.class) {
...
}

tasks.named("compileJava").dependsOn(generateSources)
```

Comment (minwoox) [reply brave/brave5/build.gradle:17]: Could you let me know if you are suggesting this convention? Or are you just asking? 😆 
If you are suggesting it, could you share the reason, please? 😉

File: brave/brave5/src/main/java/com/linecorp/armeria/client/brave/package-info.java
Comment (minwoox) [reply brave/brave5/src/main/java/com/linecorp/armeria/client/brave/package-info.java:2]: I believe this wasn't a new file because we used to have it but we moved it to another module and back here again. 😆  So let me just keep the year.

File: brave/brave5/build.gradle
Comment (kojilin) [reply brave/brave5/build.gradle:17]: Maybe both, it's good to avoid some configuration time. And saw our native-image-config using similar way.
Notice this recently https://docs.gradle.org/current/userguide/task_configuration_avoidance.html

Comment (minwoox) [reply brave/brave5/build.gradle:17]: Thank you for sharing the link. I wasn't aware that we were using an anti-pattern. 😓 Since we've employed this pattern across all our build files, let me create another PR to address this issue.
```
Avoid referencing a task by name. In the majority of cases, referencing a task by name is a fragile pattern and should be avoided. Although the task name is available on the TaskProvider, effort should be made to use references from a strongly typed model instead.
```

**Review by kojilin — APPROVED** [review]

**Comment by minwoox** [comment]
Thanks, @codefromthecrypt!

**Pull request merged**
